### ISSUE TITLE
How can I access parent source of source in resolve delegate?

### ISSUE THREAD

**Issue by johnrutherford** [graphql-dotnet/graphql-dotnet#466]
Here is a very simplified example of what I'm trying to do:

``` csharp
public class CustomerType : ObjectGraphType<Customer>
{
    public CustomerType(IOrderRepository orders)
    {
        Field(x => x.CustomerId);
        Field<OrderType>("Orders", resolve: context =>
        {
            var customer = context.Source as Customer;
            return orders.GetCustomerOrders(customer.CustomerId);
        });
    }
}

public class OrderType : ObjectGraphType<Order>
{
    public OrderType()
    {
        Field(x => x.OrderId);
        Field("Address", resolve: context =>
        {
            // There is an Address property on Customer
            // How can I get a reference to the customer?
        });
    }
}


public class AddressType : ObjectGraphType<Address>
{
    public AddressType()
    {
        Field(x => x.AddressId);
        Field(x => x.Line1)
    }
}
```

Within the resolve delegate for the Address field, how can I get a reference to the customer? This is assuming that I am only querying orders as a field of a customer, not on its own.

Thanks!

**Comment by johnrutherford** [comment]
I decided to store the customer object on the UserContext when it's loaded. Then I can access it from the resolve context on the order.

**Comment by darting** [comment]
Hi there, other than store the parent object, is there any other way? thanks

**Comment by darting** [comment]
hmm I think just
```c#
resolve : context => {
    var parent = context.Source;
}
```
?

**Comment by joemcbride** [comment]
Yes, context.Source is the object resolved from the parent field.  If on the root Query/Mutation/Subscription then the Source is the Root object from the Execute Options.

**Comment by emulicus** [comment]
Actually, I have the same question. Context.Source doesn't really work since it's only giving you the closest parent, in this case order object. Is there a way to access it's parent? What does context.RootValue represent ?

**Comment by joemcbride** [comment]
`ResolveFieldContext.RootValue` == `ExecutionOptions.Root`

Yes, `ResolveFieldContext.Source` is only one level up.  There is no built-in function to access further up the hierarchy.

**Comment by emulicus** [comment]
So, what's the best way to access the higher levels? I don't really like the idea of using UserContext for this.

**Comment by joemcbride** [comment]
@emulicus There _isn't_ a way inside this library to do it.  It would have to be an added feature.  That's not part of the GraphQL spec so its never been on the roadmap.  Resolvers can run in parallel and the parent hierarchy could be different between resolvers, so this would be a non trivial feature to add.

You could write your own context class that tried to keep track of the relationships, adding that relationship in each resolver.  You could also add an ID to the parent on your child classes, providing yourself a breadcrumb to backtrack.  Those are just a couple ideas I can think of to solve it.

**Comment by emulicus** [comment]
I see. Thank you Joe.

**Comment by sungam3r** [comment]
@emulicus @johnrutherford @darting Fixed, see #2295
